### DISCUSSION TITLE
Keycloak versioning on release

### DISCUSSION THREAD

**Discussion by NikPiermafrost** [keycloak/keycloak#17272] [Ideas]
Hi, this is my first post on this forum so i hope i get this right.
I find a bit hard to follow the current release scheme. On november 22, keycloak 20 was released. I started adopting keycloak 20.0.3 and then somewhat about 2 weeks it bumped to keycloak 20.0.5. Today, 3 months before keycloak 20, keycloak 21 is released. For what i am concerned, a major version bump seems some breaking changes are released, but i didn't see that much of a differences, so i may suggest an improvement.

- the current system for bug fixes and security issues is fine, so increasing the patch is fine
- When new non breaking features are released, bump the minor. I have not yet seen a singe X.1.X version from keycloak
- When breaking changes are introduced, bump the major

It's a bit hard currently to track when to upgrade without fearing to have to do some major work to adapt to the new version. Hopefully this is a welcome suggestion for anyone other than me and thanks for any response / feedback

**Comment by sschu** [comment]
The versioning scheme was changed quite some time ago to only have major releases, so that is intentional. In general, Keycloak tries to not have breaking changes so the differentiation of major and minor does not make that much sense. The fact that 20.0.4 is missing is due to some problem uploading artefacts to maven central, that's not the normal release procedure.

**Reply by NikPiermafrost** [reply]
Thank you for the response. I then suppose there is not a possibility to be reconsidered in the future, am i correct? It started to be pretty hard to track everything changed to check if i have to change my code accordingly or not without checking the changelog and release notes often. If the versioning system would not change in the near future, is there a method to better keep track of the changes / features / breaking changes?

**Reply by sschu** [reply]
I would suggest to add test automation around your use case, that way you immediately notice if there is a problem

**Comment by NikPiermafrost** [comment]
Good morning and sorry if i reopen this again. It is intended that keycloak 21 now has a minor increase? Just to check if my team has to adapt to this versioning scheme

**Reply by sschu** [reply]
It is intended but it is a deviation from the normal scheme. Normally, only the major and patch is increased but in rare cases it could also be minor like for 21.1.
